opentracing-impl-java8/src/main/java/io/opentracing/propagation/AbstractTracer.java

```diff
@@ -0,0 +1,175 @@
+    }
+
+    @Override
+    public SpanBuilder buildSpan(String operationName){
+        return new SpanBuilderImpl().withOperationName(operationName);
+    }
+
+    @Override
+    public <T> void inject(Span span, T carrier) {
+        registry.getInjector((Class<T>)carrier.getClass()).injectSpan(span, carrier);
+    }
+
+    @Override
+    public <T> SpanBuilder join(T carrier) {
+        return registry.getExtractor((Class<T>)carrier.getClass()).joinTrace(carrier);
+    }
+
+    public <T> Injector<T> register(Class<T> carrierType, Injector<T> injector) {
+        return registry.register(carrierType, injector);
+    }
+
+    public <T> Extractor<T> register(Class<T> carrierType, Extractor<T> extractor) {
+        return registry.register(carrierType, extractor);
+    }
+
+    private static class PropagationRegistry {
+
+        private final ConcurrentMap<Class, Injector> injectors = new ConcurrentHashMap<>();
+        private final ConcurrentMap<Class, Extractor> extractors = new ConcurrentHashMap<>();
+
+        public <T> Injector<T> getInjector(Class<T> carrierType) {
+            Class<?> c = carrierType;
+            // match first on concrete classes
+            do {
+                if (injectors.containsKey(c)) {
+                    return injectors.get(c);
+                }
+                c = c.getSuperclass();
+            } while (c != null);
+            // match second on interfaces
+            for (Class<?> iface : carrierType.getInterfaces()) {
+                if (injectors.containsKey(c)) {
+                    return injectors.get(c);
+                }
+            }
+            throw new AssertionError("no registered injector for " + carrierType.getName());
+        }
+
+        public <T> Extractor<T> getExtractor(Class<T> carrierType) {
+            Class<?> c = carrierType;
+            // match first on concrete classes
+            do {
+                if (extractors.containsKey(c)) {
+                    return extractors.get(c);
+                }
+                c = c.getSuperclass();
+            } while (c != null);
+            // match second on interfaces
+            for (Class<?> iface : carrierType.getInterfaces()) {
+                if (extractors.containsKey(c)) {
+                    return extractors.get(c);
+                }
+            }
+            throw new AssertionError("no registered extractor for " + carrierType.getName());
+        }
+
+        public <T> Injector<T> register(Class<T> carrierType, Injector<T> injector) {
+            return injectors.putIfAbsent(carrierType, injector);
+        }
+
+        public <T> Extractor<T> register(Class<T> carrierType, Extractor<T> extractor) {
+            return extractors.putIfAbsent(carrierType, extractor);
+        }
+    }
+
+    static class SpanBuilderImpl implements SpanBuilder {
+
+        SpanBuilderImpl() {}
+
+        final Map<String,String> traceState = new HashMap<>();
```

Comment:
how about making this into an abstract type instead?  Forcing everyone to convert trace state into strings seems like unnecessary overhead.